### PR TITLE
chore(sdk): merge v2 dsl code back to v1.

### PR DESCRIPTION
**Pull request by chensun**
**Description of your changes:**

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: sdk/python/kfp/dsl/_container_op.py [hunk 1]
```diff
@@ -1046,8 +1078,15 @@ def __init__(
         sidecars=sidecars,
         is_exit_handler=is_exit_handler)
 
-    if not ContainerOp._DISABLE_REUSABLE_COMPONENT_WARNING and '--component_launcher_class_path' not in (
-        arguments or []):
+    # Check if the pipeline is being compiled for v2 spec.
+    self.is_compiling_for_v2 = False
```
Comment (neuromage) [inline sdk/python/kfp/dsl/_container_op.py:1082]: Instead of doing this, assume compiling_for_v2 is always True and remove this altogether. The only downside is the resource spec for gpus below, which we can validate when submitting to the backend.

Comment (chensun) [reply sdk/python/kfp/dsl/_container_op.py]: Done

File: sdk/python/kfp/dsl/_component_bridge.py [hunk 2]
```diff
@@ -124,4 +142,251 @@ def _create_container_op_from_component_and_arguments(
...
+          command=resolved_cmd.command,
+          args=resolved_cmd.args))
+
+  # Override command and arguments if compiling to v2.
+  if is_compiling_for_v2:
+    task.command = resolved_cmd.command
+    task.arguments = resolved_cmd.args
+
+    # limit this to v2 compiling only to avoid possible behavior change in v1.
+    task.inputs = input_params
```
Comment (chensun) [inline sdk/python/kfp/dsl/_component_bridge.py:392]: @neuromage, I will attempt to remove the above `if is_component_for_v2:` branch in a separate PR. There're some weird issues if I remove this, yet I haven't figured out the best way to fix it. So just keep it for now, so it reserves the current behaviors for both v1 and v2 compiling.

TL;DR
The symptom of removing this branch is when I compile the custom job component (https://github.com/kubeflow/pipelines/pull/5146), `task.command`, if not override by the v2 resolved command, will contain `{{pipelineparam:op=;name=pipeline-output-directory}}` as a result of https://github.com/kubeflow/pipelines/pull/5122.

and `task.inputs` is empty, so code below 
(specifically `self._inputs += _pipeline_param.extract_pipelineparams_from_any(...)`) will recognize there's a `PipelineParam` from the command and add it to the inputs. 
https://github.com/kubeflow/pipelines/blob/965bc62fd60e94de9d26603c1c06e60d931906db/sdk/python/kfp/dsl/_container_op.py#L769-L785

Having this extra input (`pipeline-output-directory`) then results an additional input in `component_spec.input_definitions`, which is wrong for v2 compilation - we don't need the pipeline root appear in the pipeline input in v2.

Comment (neuromage) [reply sdk/python/kfp/dsl/_component_bridge.py:392]: I see, ok, thanks for explaining. We'll fix in follow up PR then.
